pallets/stateful-storage/src/lib.rs

```diff
@@ -170,15 +170,37 @@ pub mod pallet {
 
 		/// Invalid item action
 		InvalidItemAction,
+
+		/// Target page hash does not match current page hash
+		StalePageState,
 	}
 
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
 	pub enum Event<T: Config> {
-		ItemizedPageUpdated { msa_id: MessageSourceId, schema_id: SchemaId },
-		ItemizedPageDeleted { msa_id: MessageSourceId, schema_id: SchemaId },
-		PaginatedPageUpdated { msa_id: MessageSourceId, schema_id: SchemaId, page_id: PageId },
-		PaginatedPageDeleted { msa_id: MessageSourceId, schema_id: SchemaId, page_id: PageId },
+		ItemizedPageUpdated {
+			msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			prev_content_hash: PageHash,
+			curr_content_hash: PageHash,
+		},
+		ItemizedPageDeleted {
+			msa_id: MessageSourceId,
+			schema_id: SchemaId,
+		},
+		PaginatedPageUpdated {
+			msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+			prev_content_hash: PageHash,
+			curr_content_hash: PageHash,
+		},
+		PaginatedPageDeleted {
+			msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+			prev_content_hash: PageHash,
```

Comment:
nit: To be consistent we should either include `prev_content_hash` for both `ItemizedPageDeleted` and `PaginatedPageDeleted` or for none of them.

Reply:
Fair point. Done.